libraries/standard/mqtt/utest/mqtt_lightweight_utest.c

```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
     return ( size_t ) ( pBuffer - pDestination );
 }
 
+/* ========================================================================== */
+
+/**
+ * @brief Tests that MQTT_GetConnectPacketSize works as intended.
+ */
+void test_MQTT_GetConnectPacketSize( void )
+{
+    MQTTConnectInfo_t connectInfo;
+    size_t remainingLength = 0;
+    size_t packetSize = 0;
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishInfo_t willInfo = { 0 };
+
+    /* Call MQTT_GetConnectPacketSize() with various combinations of
+     * incorrect paramters */
+
+    status = MQTT_GetConnectPacketSize( NULL, NULL, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    status = MQTT_GetConnectPacketSize( &connectInfo, NULL, NULL, &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    status = MQTT_GetConnectPacketSize( &connectInfo, NULL, &remainingLength, NULL );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    /* Verify empty connect info fails. */
+    memset( ( void * ) &connectInfo, 0x0, sizeof( connectInfo ) );
+    status = MQTT_GetConnectPacketSize( &connectInfo, NULL, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    /* Verify empty client identifier fails. */
+    connectInfo.pClientIdentifier = CLIENT_IDENTIFIER;
+    connectInfo.clientIdentifierLength = 0;
+    status = MQTT_GetConnectPacketSize( &connectInfo, NULL, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    connectInfo.pClientIdentifier = NULL;
+    connectInfo.clientIdentifierLength = CLIENT_IDENTIFIER_LENGTH;
+    status = MQTT_GetConnectPacketSize( &connectInfo, NULL, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    /* Verify good case */
+    memset( ( void * ) &connectInfo, 0x0, sizeof( connectInfo ) );
+    connectInfo.cleanSession = true;
+    connectInfo.pClientIdentifier = "TEST";
+    connectInfo.clientIdentifierLength = 4;
+    status = MQTT_GetConnectPacketSize( &connectInfo, NULL, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
+    /* Make sure remaining size returned is 16. */
+    TEST_ASSERT_EQUAL_INT( 16, remainingLength );
+    /* Make sure packet size is 18. */
+    TEST_ASSERT_EQUAL_INT( 18, packetSize );
+
+    /* With will. These parameters will cause the packet to be
+     * 4 + 2 + 8 + 2 = 16 bytes larger. */
+    willInfo.pTopicName = "test";
+    willInfo.topicNameLength = 4;
+    willInfo.pPayload = "testload";
+    willInfo.payloadLength = 8;
+    status = MQTT_GetConnectPacketSize( &connectInfo, &willInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
+    /* Make sure remaining size returned is 32 = 16 + 16. */
+    TEST_ASSERT_EQUAL_INT( 32, remainingLength );
+    /* Make sure packet size is 34 = 18 + 16. */
+    TEST_ASSERT_EQUAL_INT( 34, packetSize );
+}
```

Comment:
I see that the paths for pUserName != NULL and pPassword != NULL and connectPacketSize > MQTT_PACKET_CONNECT_MAX_SIZE are covered in the unit tests for MQTT_SerializeConnect. Considering MQTT_GetConnectPacketSize as a unit, I think those cases belong here.

Reply:
Added